src/parser.cpp

```diff
@@ -106,6 +120,273 @@ Buffer Usage
 */
 //-----------------------------------------------
 
+namespace {
+
+struct discontiguous_iterator
+{
+    buffers::const_buffer const* pos = nullptr;
+    std::size_t off = 0;
+
+    discontiguous_iterator(
+        buffers::const_buffer const* pos_)
+    : pos(pos_)
+    {
+    }
+
+    char
+    operator*() const noexcept
+    {
+        BOOST_ASSERT(pos);
+        auto it =
+            reinterpret_cast<char const*>(pos->data()) +
+            off;
+        return *it;
+    }
+
+    discontiguous_iterator&
+    operator++() noexcept
+    {
+        if( off >= pos->size() )
+        {
+            ++pos;
+            off = 0;
+            return *this;
+        }
+        ++off;
+        return *this;
+    }
+
+    discontiguous_iterator
+    operator++(int) noexcept
+    {
+        auto old = *this;
+        ++*this;
+        return old;
+    }
+
+    bool
+    operator==(
+        discontiguous_iterator const& rhs) const noexcept
+    {
+        return pos == rhs.pos;
+    }
+
+    bool
+    operator!=(
+        discontiguous_iterator const& rhs) const noexcept
+    {
+        return !(*this == rhs);
+    }
+};
+
+constexpr static
+std::size_t const max_chunk_header_len = 16 + 2;
+
+constexpr static
+std::size_t const last_chunk_len = 5;
+
+void
+parse_chunk_header(
```

Comment:
should this function be `static` ?

Reply:
I had these functions living originally in an unnamed namespace but this isn't readable. I'll update the code to eschew the unnamed namespace and just use static like normal.